File: examples/go/telemetry.go
// Copyright 2026 Chengxi Luo
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package main

import (
	"encoding/binary"
	"math"
	"math/big"
)

// TelemetryFrame mirrors examples/cases/telemetry.yaml — one reading from a
// field device.
//
// This type is what exercises the corners the other examples do not: a
// `uint128` address, two differently shaped fixed arrays, an `optional<float32>`
// (the only `optional<scalar>` in the suite), a `map<string,uint64>`, and float
// cases covering NaN, ±Inf and negative zero. Its `binary` format is the only
// one declared, because NaN and Inf have no JSON representation.
//
// Binary layout follows the conventions in wire.go, with two additions:
//
//	uint128      16 bytes, little-endian
//	array<T,N>   exactly N elements, no count prefix (N is fixed by the schema)
type TelemetryFrame struct {
	DeviceID      uint64            `serify:"device_id"`
	IPv6          *big.Int          `serify:"ipv6"`
	LocalIP       [4]uint8          `serify:"local_ip"`
	Firmware      string            `serify:"firmware"`
	BootCount     uint16            `serify:"boot_count"`
	RSSIdBm       int8              `serify:"rssi_dbm"`
	TemperatureDC int16             `serify:"temperature_dc"`
	ClockDriftMS  int32             `serify:"clock_drift_ms"`
	BatteryVolts  float32           `serify:"battery_volts"`
	Latitude      float64           `serify:"latitude"`
	Longitude     float64           `serify:"longitude"`
	HumidityPct   *float32          `serify:"humidity_pct"`
	AccelMg       [3]int16          `serify:"accel_mg"`
	VisibleCells  []uint32          `serify:"visible_cells"`
	PacketCounts  map[string]uint64 `serify:"packet_counts"`
	GPSFix        bool              `serify:"gps_fix"`
	Signature     []byte            `serify:"signature"`
}

func (t *TelemetryFrame) MarshalBinary() ([]byte, error) {
	buf := binary.LittleEndian.AppendUint64(nil, t.DeviceID)

	// uint128 is unsigned, so the same 16 two's-complement bytes ledger.go uses
	// for int128 serve here without a sign to re-apply on the way back.
	buf = appendInt128(buf, t.IPv6)

	buf = append(buf, t.LocalIP[:]...)
	buf = appendLenStr(buf, t.Firmware)
	buf = binary.LittleEndian.AppendUint16(buf, t.BootCount)
	buf = append(buf, byte(t.RSSIdBm))
	buf = binary.LittleEndian.AppendUint16(buf, uint16(t.TemperatureDC))
	buf = binary.LittleEndian.AppendUint32(buf, uint32(t.ClockDriftMS))
	buf = binary.LittleEndian.AppendUint32(buf, math.Float32bits(t.BatteryVolts))
	buf = binary.LittleEndian.AppendUint64(buf, math.Float64bits(t.Latitude))
	buf = binary.LittleEndian.AppendUint64(buf, math.Float64bits(t.Longitude))

	// optional<float32>: a presence flag, then the value if present.
	if t.HumidityPct == nil {
		buf = append(buf, 0)
	} else {
		buf = append(buf, 1)
		buf = binary.LittleEndian.AppendUint32(buf, math.Float32bits(*t.HumidityPct))
	}

	for _, v := range t.AccelMg {
		buf = binary.LittleEndian.AppendUint16(buf, uint16(v))
	}

	buf = appendCount(buf, t.VisibleCells)
	for _, v := range t.VisibleCells {
		buf = binary.LittleEndian.AppendUint32(buf, v)
	}

	// Map keys are sorted so the bytes are deterministic; see wire.go.
	keys := sortedKeys(t.PacketCounts)
	buf = binary.LittleEndian.AppendUint32(buf, uint32(len(keys)))
	for _, k := range keys {
		buf = appendLenStr(buf, k)
		buf = binary.LittleEndian.AppendUint64(buf, t.PacketCounts[k])
	}

	if t.GPSFix {
		buf = append(buf, 1)
	} else {
		buf = append(buf, 0)
	}

	buf = binary.LittleEndian.AppendUint32(buf, uint32(len(t.Signature)))
	buf = append(buf, t.Signature...)

	return buf, nil
}

//nolint:gocognit,funlen // one read block per field
func (t *TelemetryFrame) UnmarshalBinary(b []byte) error {
	if len(b) < 8 {
		return errTruncated
	}
	t.DeviceID = binary.LittleEndian.Uint64(b)
	b = b[8:]

	var err error
	if t.IPv6, b, err = readInt128(b); err != nil {
		return err
	}
	// readInt128 re-applies a sign; uint128 has none, so map a negative back
	// into the unsigned residue class it came from.
	if t.IPv6.Sign() < 0 {
		t.IPv6 = new(big.Int).Add(t.IPv6, new(big.Int).Lsh(big.NewInt(1), 128))
	}

	if len(b) < len(t.LocalIP) {
		return errTruncated
	}
	copy(t.LocalIP[:], b[:len(t.LocalIP)])
	b = b[len(t.LocalIP):]

	if t.Firmware, b, err = readLenStr(b); err != nil {
		return err
	}

	if len(b) < 2+1+2+4+4+8+8+1 {
		return errTruncated
	}
	t.BootCount = binary.LittleEndian.Uint16(b)
	t.RSSIdBm = int8(b[2])
	t.TemperatureDC = int16(binary.LittleEndian.Uint16(b[3:]))
	t.ClockDriftMS = int32(binary.LittleEndian.Uint32(b[5:]))
	t.BatteryVolts = math.Float32frombits(binary.LittleEndian.Uint32(b[9:]))
	t.Latitude = math.Float64frombits(binary.LittleEndian.Uint64(b[13:]))
	t.Longitude = math.Float64frombits(binary.LittleEndian.Uint64(b[21:]))
	present := b[29]
	b = b[30:]

	t.HumidityPct = nil
	if present != 0 {
		if len(b) < 4 {
			return errTruncated
		}
		v := math.Float32frombits(binary.LittleEndian.Uint32(b))
		t.HumidityPct = &v
		b = b[4:]
	}

	if len(b) < len(t.AccelMg)*2 {
		return errTruncated
	}
	for i := range t.AccelMg {
		t.AccelMg[i] = int16(binary.LittleEndian.Uint16(b[i*2:]))
	}
	b = b[len(t.AccelMg)*2:]

	n, b, err := readCount(b)
	if err != nil {
		return err
	}
	if len(b) < n*4 {
		return errTruncated
	}
	t.VisibleCells = make([]uint32, n)
	for i := range n {
		t.VisibleCells[i] = binary.LittleEndian.Uint32(b[i*4:])
	}
	b = b[n*4:]

	if n, b, err = readCount(b); err != nil {
		return err
	}
	t.PacketCounts = make(map[string]uint64, n)
	for range n {
		var k string
		if k, b, err = readLenStr(b); err != nil {
			return err
		}
		if len(b) < 8 {
			return errTruncated
		}
		t.PacketCounts[k] = binary.LittleEndian.Uint64(b)
		b = b[8:]
	}

	if len(b) < 1 {
		return errTruncated
	}
	t.GPSFix = b[0] != 0
	b = b[1:]

	if n, b, err = readCount(b); err != nil {
		return err
	}
	if len(b) < n {
		return errTruncated
	}
	t.Signature = append([]byte(nil), b[:n]...)

	return nil
}

File: examples/rust/src/telemetry.rs
// Copyright 2026 Chengxi Luo
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

//! `TelemetryFrame` mirrors examples/cases/telemetry.yaml — one reading from a
//! field device.
//!
//! This is the type that could not be written until recently: its `humidity_pct`
//! is an `optional<float32>`, and the derive recognised only `Option<String>`, so
//! `Option<f32>` failed to compile. Between them its fields also cover the
//! suite's only `uint128`, two differently shaped fixed arrays, a
//! `map<string,uint64>`, and float values including NaN, ±Inf and negative zero.
//!
//! Go is the --ref language and owns the byte layout; see examples/go/wire.go.

use std::collections::HashMap;

use serify::SerifyModel;

use crate::wire::{append_len_str, read_bytes, read_len_str};

#[derive(SerifyModel)]
pub struct TelemetryFrame {
    device_id: u64,
    ipv6: u128,
    local_ip: [u8; 4],
    firmware: String,
    boot_count: u16,
    rssi_dbm: i8,
    temperature_dc: i16,
    clock_drift_ms: i32,
    battery_volts: f32,
    latitude: f64,
    longitude: f64,
    humidity_pct: Option<f32>,
    accel_mg: [i16; 3],
    visible_cells: Vec<u32>,
    packet_counts: HashMap<String, u64>,
    gps_fix: bool,
    signature: Vec<u8>,
}

impl TelemetryFrame {
    pub fn marshal(&self) -> Result<Vec<u8>, String> {
        let mut buf = Vec::new();
        buf.extend_from_slice(&self.device_id.to_le_bytes());
        buf.extend_from_slice(&self.ipv6.to_le_bytes());
        buf.extend_from_slice(&self.local_ip);
        append_len_str(&mut buf, &self.firmware);
        buf.extend_from_slice(&self.boot_count.to_le_bytes());
        buf.extend_from_slice(&self.rssi_dbm.to_le_bytes());
        buf.extend_from_slice(&self.temperature_dc.to_le_bytes());
        buf.extend_from_slice(&self.clock_drift_ms.to_le_bytes());
        buf.extend_from_slice(&self.battery_volts.to_le_bytes());
        buf.extend_from_slice(&self.latitude.to_le_bytes());
        buf.extend_from_slice(&self.longitude.to_le_bytes());

        // optional<float32>: a presence flag, then the value if present.
        match self.humidity_pct {
            None => buf.push(0),
            Some(v) => {
                buf.push(1);
                buf.extend_from_slice(&v.to_le_bytes());
            }
        }

        for v in self.accel_mg.iter() {
            buf.extend_from_slice(&v.to_le_bytes());
        }

        buf.extend_from_slice(&(self.visible_cells.len() as u32).to_le_bytes());
        for v in self.visible_cells.iter() {
            buf.extend_from_slice(&v.to_le_bytes());
        }

        let keys: Vec<&String> = self.packet_counts.keys().collect();
        buf.extend_from_slice(&(keys.len() as u32).to_le_bytes());
        for k in keys {
            append_len_str(&mut buf, k);
            buf.extend_from_slice(&self.packet_counts[k].to_le_bytes());
        }

        buf.push(self.gps_fix as u8);
        buf.extend_from_slice(&(self.signature.len() as u32).to_le_bytes());
        buf.extend_from_slice(&self.signature);

        Ok(buf)
    }

    pub fn unmarshal(data: &[u8]) -> Result<Self, String> {
        let mut p = 0usize;

        let device_id = u64::from_le_bytes(read_bytes!(data, p, 8).try_into().unwrap());
        let ipv6 = u128::from_le_bytes(read_bytes!(data, p, 16).try_into().unwrap());
        let mut local_ip = [0u8; 4];
        local_ip.copy_from_slice(read_bytes!(data, p, 4));
        let firmware = read_len_str(data, &mut p)?;
        let boot_count = u16::from_le_bytes(read_bytes!(data, p, 2).try_into().unwrap());
        let rssi_dbm = i8::from_le_bytes(read_bytes!(data, p, 1).try_into().unwrap());
        let temperature_dc = i16::from_le_bytes(read_bytes!(data, p, 2).try_into().unwrap());
        let clock_drift_ms = i32::from_le_bytes(read_bytes!(data, p, 4).try_into().unwrap());
        let battery_volts = f32::from_le_bytes(read_bytes!(data, p, 4).try_into().unwrap());
        let latitude = f64::from_le_bytes(read_bytes!(data, p, 8).try_into().unwrap());
        let longitude = f64::from_le_bytes(read_bytes!(data, p, 8).try_into().unwrap());

        let humidity_pct = if read_bytes!(data, p, 1)[0] == 0 {
            None
        } else {
            Some(f32::from_le_bytes(
                read_bytes!(data, p, 4).try_into().unwrap(),
            ))
        };

        let mut accel_mg = [0i16; 3];
        for a in accel_mg.iter_mut() {
            *a = i16::from_le_bytes(read_bytes!(data, p, 2).try_into().unwrap());
        }

        let cell_count = u32::from_le_bytes(read_bytes!(data, p, 4).try_into().unwrap()) as usize;
        let mut visible_cells = Vec::with_capacity(cell_count);
        for _ in 0..cell_count {
            visible_cells.push(u32::from_le_bytes(
                read_bytes!(data, p, 4).try_into().unwrap(),
            ));
        }

        let entry_count = u32::from_le_bytes(read_bytes!(data, p, 4).try_into().unwrap()) as usize;
        let mut packet_counts = HashMap::with_capacity(entry_count);
        for _ in 0..entry_count {
            let k = read_len_str(data, &mut p)?;
            let v = u64::from_le_bytes(read_bytes!(data, p, 8).try_into().unwrap());
            packet_counts.insert(k, v);
        }

        let gps_fix = read_bytes!(data, p, 1)[0] != 0;
        let sig_len = u32::from_le_bytes(read_bytes!(data, p, 4).try_into().unwrap()) as usize;
        let signature = read_bytes!(data, p, sig_len).to_vec();
        let _ = p; // read_bytes! advances the cursor; this is the last field.

        Ok(TelemetryFrame {
            device_id,
            ipv6,
            local_ip,
            firmware,
            boot_count,
            rssi_dbm,
            temperature_dc,
            clock_drift_ms,
            battery_volts,
            latitude,
            longitude,
            humidity_pct,
            accel_mg,
            visible_cells,
            packet_counts,
            gps_fix,
            signature,
        })
    }
}
